File: phases/09-computer-networks/20-software-defined-networking-ebpf/code/main.py
"""
Software-Defined Networking & eBPF
Phase 09 — Computer Networks

An SDN controller simulation with OpenFlow-like flow tables,
a learning switch, and a userspace eBPF/XDP packet counter.
"""


class FlowRule:
    """An OpenFlow-style flow table entry with match fields and actions."""

    def __init__(self, priority=0, match=None, actions=None, idle_timeout=0):
        self.priority = priority
        self.match = match or {}
        self.actions = actions or []
        self.idle_timeout = idle_timeout
        self.packet_count = 0
        self.byte_count = 0

    def matches(self, packet):
        """Return True if this flow rule matches the given packet."""
        for key, value in self.match.items():
            if packet.get(key) != value:
                return False
        return True

    def apply_actions(self, packet):
        """Execute actions for this rule; update counters."""
        self.packet_count += 1
        self.byte_count += packet.get("length", 0)
        return self.actions


class FlowTable:
    """A switch's flow table with priority-ordered rules."""

    def __init__(self):
        self.rules = []

    def add_rule(self, rule):
        self.rules.append(rule)
        self.rules.sort(key=lambda r: r.priority, reverse=True)

    def remove_rule(self, match):
        self.rules = [r for r in self.rules if r.match != match]

    def lookup(self, packet):
        """Return actions for highest-priority matching rule, or None."""
        for rule in self.rules:
            if rule.matches(packet):
                return rule.apply_actions(packet)
        return None


class OpenFlowSwitch:
    """An OpenFlow-capable switch with flow tables and controller connection."""

    def __init__(self, switch_id, controller=None):
        self.switch_id = switch_id
        self.flow_table = FlowTable()
        self.controller = controller
        self.ports = {}

    def add_port(self, port_id, name=""):
        self.ports[port_id] = {"name": name, "stats": {"rx_packets": 0, "tx_packets": 0}}

    def receive_packet(self, packet):
        """Process an incoming packet: flow lookup -> execute or PACKET_IN."""
        actions = self.flow_table.lookup(packet)
        if actions:
            return self._execute_actions(packet, actions)
        if self.controller:
            self.controller.handle_packet_in(self, packet)
            return ["PACKET_IN"]
        return []

    def _execute_actions(self, packet, actions):
        outputs = []
        for action in actions:
            if action == "flood":
                outputs = list(self.ports.keys())
            elif action == "drop":
                return ["drop"]
            elif action.startswith("output:"):
                port = int(action.split(":")[1])
                if port in self.ports:
                    outputs.append(port)
        for port in outputs:
            if port in self.ports:
                self.ports[port]["stats"]["tx_packets"] += 1
        return outputs


class SDNController:
    """An SDN controller managing switches and installing flow rules."""

    def __init__(self):
        self.switches = {}
        self._mac_table = {}

    def add_switch(self, sw):
        sw.controller = self
        self.switches[sw.switch_id] = sw

    def handle_packet_in(self, switch, packet):
        """Handle PACKET_IN: learn MAC, install flow rule for known dst, or flood."""
        src_mac = packet.get("src_mac")
        in_port = packet.get("in_port")
        self._mac_table[src_mac] = (switch.switch_id, in_port)

        dst_mac = packet.get("dst_mac")
        entry = self._mac_table.get(dst_mac)
        if entry is not None:
            _, out_port = entry
            rule = FlowRule(
                priority=100,
                match={"dst_mac": dst_mac, "in_port": in_port},
                actions=[f"output:{out_port}"],
            )
            switch.flow_table.add_rule(rule)
        else:
            rule = FlowRule(
                priority=10,
                match={"dst_mac": dst_mac},
                actions=["flood"],
            )
            switch.flow_table.add_rule(rule)

    def install_flow(self, switch_id, match, actions, priority=100):
        """Install a flow rule via the OpenFlow protocol simulation."""
        if switch_id in self.switches:
            rule = FlowRule(priority=priority, match=match, actions=actions)
            self.switches[switch_id].flow_table.add_rule(rule)

    def get_stats(self):
        """Return per-switch statistics."""
        stats = {}
        for sid, sw in self.switches.items():
            tx = sum(p["stats"]["tx_packets"] for p in sw.ports.values())
            stats[sid] = {
                "ports": {pid: p["stats"] for pid, p in sw.ports.items()},
                "flow_rules": len(sw.flow_table.rules),
                "packets_forwarded": tx,
            }
        return stats


class BPFPacketCounter:
    """Userspace simulation of an eBPF packet counter with BPF_MAP_TYPE_ARRAY."""

    def __init__(self):
        self.protocol_counts = {"TCP": 0, "UDP": 0, "ICMP": 0, "OTHER": 0}
        self.flow_stats = {}

    def parse_packet(self, raw_data):
        """Parse Ethernet/IP/transport headers from raw bytes (simulated eBPF parser)."""
        if len(raw_data) < 14:
            return None

        eth_type = int.from_bytes(raw_data[12:14], "big")
        if eth_type != 0x0800:
            return {"eth_type": eth_type, "protocol": "OTHER"}

        if len(raw_data) < 34:
            return None

        version_ihl = raw_data[14]
        ihl = (version_ihl & 0x0F) * 4
        if ihl < 20 or len(raw_data) < 14 + ihl:
            return None

        protocol = raw_data[23]
        src_ip = ".".join(str(b) for b in raw_data[26:30])
        dst_ip = ".".join(str(b) for b in raw_data[30:34])
        total_length = int.from_bytes(raw_data[16:18], "big")

        proto_map = {6: "TCP", 17: "UDP", 1: "ICMP"}
        proto_name = proto_map.get(protocol, "OTHER")

        result = {
            "eth_type": eth_type,
            "src_ip": src_ip,
            "dst_ip": dst_ip,
            "protocol": proto_name,
            "ip_header_len": ihl,
            "total_length": total_length,
        }

        if protocol in (6, 17) and len(raw_data) >= 14 + ihl + 4:
            offset = 14 + ihl
            result["src_port"] = int.from_bytes(raw_data[offset : offset + 2], "big")
            result["dst_port"] = int.from_bytes(raw_data[offset + 2 : offset + 4], "big")

        return result

    def process(self, parsed):
        """Update BPF map counters (simulated XDP action)."""
        if parsed is None:
            return "XDP_DROP"

        proto = parsed.get("protocol", "OTHER")
        self.protocol_counts[proto] = self.protocol_counts.get(proto, 0) + 1

        flow_key = (
            parsed.get("src_ip"),
            parsed.get("dst_ip"),
            proto,
            parsed.get("src_port", 0),
            parsed.get("dst_port", 0),
        )
        self.flow_stats[flow_key] = (
            self.flow_stats.get(flow_key, 0) + parsed.get("total_length", 0)
        )

        return "XDP_PASS"

    def get_protocol_stats(self):
        return dict(self.protocol_counts)

    def get_top_flows(self, n=5):
        sorted_flows = sorted(self.flow_stats.items(), key=lambda x: x[1], reverse=True)
        return sorted_flows[:n]


def _build_test_packets():
    """Build a list of raw byte-array test packets."""
    pkts = []

    def make_pkt(proto, src_ip, dst_ip, src_port=0, dst_port=0, extra_len=0):
        p = bytearray(64)
        p[12:14] = (0x0800).to_bytes(2, "big")
        p[14] = 0x45
        length = 40 + extra_len
        p[16:18] = length.to_bytes(2, "big")
        p[23] = proto
        for i, b in enumerate(src_ip.split(".")):
            p[26 + i] = int(b)
        for i, b in enumerate(dst_ip.split(".")):
            p[30 + i] = int(b)
        if src_port:
            p[34:36] = src_port.to_bytes(2, "big")
        if dst_port:
            p[36:38] = dst_port.to_bytes(2, "big")
        return bytes(p)

    # 20 TCP packets from various hosts to 10.0.0.1:5000
    for i in range(20):
        pkts.append(make_pkt(6, f"192.168.1.{i % 5 + 1}", "10.0.0.1", 80, 5000))

    # 10 UDP DNS packets
    for _ in range(10):
        pkts.append(make_pkt(17, "192.168.1.1", "8.8.8.8", 53, 12345))

    # 5 ICMP pings
    for _ in range(5):
        pkts.append(make_pkt(1, "192.168.1.1", "8.8.8.8"))

    # 3 IPv6 packets (non-IPv4, should be OTHER)
    for _ in range(3):
        p = bytearray(64)
        p[12:14] = (0x86DD).to_bytes(2, "big")
        pkts.append(bytes(p))

    return pkts


def simulate_topology():
    """Create a 3-switch topology with learning switch behavior."""
    ctrl = SDNController()

    s1 = OpenFlowSwitch("switch-1")
    s1.add_port(1, "host-a")
    s1.add_port(2, "switch-2")
    s1.add_port(3, "switch-3")

    s2 = OpenFlowSwitch("switch-2")
    s2.add_port(1, "switch-1")
    s2.add_port(2, "host-b")

    s3 = OpenFlowSwitch("switch-3")
    s3.add_port(1, "switch-1")
    s3.add_port(2, "host-c")

    ctrl.add_switch(s1)
    ctrl.add_switch(s2)
    ctrl.add_switch(s3)

    for sw in [s1, s2, s3]:
        sw.flow_table.add_rule(FlowRule(priority=0, match={}, actions=["flood"]))

    print("=== SDN Controller Simulation ===")
    print()
    print("Topology: host-a -- s1 -- s2 -- host-b")
    print("                          \\")
    print("                           s3 -- host-c")
    print()

    # First packet — table miss triggers PACKET_IN
    print("1. Host A sends ARP (first packet — table miss)")
    p1 = {
        "src_mac": "00:11:22:33:44:AA",
        "dst_mac": "FF:FF:FF:FF:FF:FF",
        "in_port": 1,
        "eth_type": 0x0806,
        "length": 64,
    }
    result = s1.receive_packet(p1)
    print(f"   Switch s1 action: {result}")
    print(f"   Controller learns MAC 00:11:22:33:44:AA on s1 port 1")

    # Install static flow rule
    print()
    print("2. Controller installs static flow rule: host A traffic -> port 2")
    ctrl.install_flow(
        "switch-1",
        match={"src_mac": "00:11:22:33:44:AA"},
        actions=["output:2"],
        priority=200,
    )
    print("   Flow rule installed")

    # Second packet hits the flow table directly
    print()
    print("3. Host A sends IPv4 — flow table hit (fast path)")
    p2 = {
        "src_mac": "00:11:22:33:44:AA",
        "dst_mac": "00:11:22:33:44:BB",
        "in_port": 1,
        "eth_type": 0x0800,
        "length": 512,
    }
    result = s1.receive_packet(p2)
    print(f"   Switch s1 action: {result}")

    # Stats
    print()
    print("4. Switch statistics:")
    stats = ctrl.get_stats()
    for sid, data in stats.items():
        print(f"   {sid}: {data['packets_forwarded']} pkts fwd, "
              f"{data['flow_rules']} flow rules")

    print()
    print("5. Flow rule traffic (packet counts):")
    for rule in s1.flow_table.rules:
        print(f"   Pri {rule.priority}: match={rule.match}, "
              f"count={rule.packet_count} pkts")


def simulate_ebpf():
    """Simulate eBPF XDP packet processing with protocol counting."""
    print()
    print("=== eBPF Packet Counter (XDP Simulation) ===")

    counter = BPFPacketCounter()
    test_packets = _build_test_packets()

    print()
    print(f"Processing {len(test_packets)} packets through simulated eBPF/XDP...")

    for pkt in test_packets:
        parsed = counter.parse_packet(pkt)
        counter.process(parsed)

    print()
    print("Protocol distribution (simulated BPF_MAP_TYPE_ARRAY):")
    for proto, count in counter.get_protocol_stats().items():
        bar = "#" * count
        print(f"  {proto:6s}: {count:3d} packets {bar}")

    print()
    print("Top flows by byte volume:")
    for (src, dst, proto, sp, dp), bytes_n in counter.get_top_flows(5):
        print(f"  {src}:{sp} -> {dst}:{dp} ({proto}): {bytes_n} bytes")


def main():
    simulate_topology()
    simulate_ebpf()


if __name__ == "__main__":
    main()

File: phases/09-computer-networks/20-software-defined-networking-ebpf/code/main.c
/**
 * Software-Defined Networking & eBPF
 * Phase 09 — Computer Networks
 *
 * Userspace simulation of an eBPF XDP packet counter.
 * Parses Ethernet/IP/transport headers, counts packets by protocol
 * using simulated BPF_MAP_TYPE_ARRAY, and tracks per-flow byte counts.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

/* ------------------------------------------------------------------ */
/*  Byte-order helpers (avoid <arpa/inet.h> dependency)               */
/* ------------------------------------------------------------------ */

static inline uint16_t ntoh16(uint16_t x) {
    return ((x & 0xFF) << 8) | ((x >> 8) & 0xFF);
}

static inline uint32_t ntoh32(uint32_t x) {
    return ((x & 0xFF) << 24) | ((x & 0xFF00) << 8) |
           ((x >> 8) & 0xFF00) | ((x >> 24) & 0xFF);
}

/* ------------------------------------------------------------------ */
/*  Simulated packet header structures (packed to match wire format)  */
/* ------------------------------------------------------------------ */

struct eth_hdr {
    uint8_t  dst_mac[6];
    uint8_t  src_mac[6];
    uint16_t eth_type;
} __attribute__((packed));

struct ipv4_hdr {
    uint8_t  version_ihl;
    uint8_t  dscp_ecn;
    uint16_t total_length;
    uint16_t identification;
    uint16_t flags_fragment;
    uint8_t  ttl;
    uint8_t  protocol;
    uint16_t checksum;
    uint32_t src_ip;
    uint32_t dst_ip;
} __attribute__((packed));

/* ------------------------------------------------------------------ */
/*  Simulated BPF maps                                                */
/* ------------------------------------------------------------------ */

#define MAX_PROTO 4
static const char *proto_names[MAX_PROTO] = {"TCP", "UDP", "ICMP", "OTHER"};
static uint64_t proto_counts[MAX_PROTO] = {0};

#define MAX_FLOWS 256
typedef struct {
    uint32_t src_ip;
    uint32_t dst_ip;
    uint8_t  protocol;
    uint16_t src_port;
    uint16_t dst_port;
    uint64_t bytes;
} flow_t;
static flow_t flows[MAX_FLOWS];
static int    flow_cnt = 0;

/* ------------------------------------------------------------------ */
/*  Simulated XDP actions                                             */
/* ------------------------------------------------------------------ */

#define XDP_PASS 1
#define XDP_DROP 0

/* ------------------------------------------------------------------ */
/*  XDP program — mirrors a real eBPF/XDP program that:              */
/*   – parses Ethernet + IPv4 + TCP/UDP/ICMP headers                 */
/*   – increments per-protocol counters (BPF_MAP_TYPE_ARRAY)         */
/*   – updates per-flow byte counters (BPF_MAP_TYPE_HASH)            */
/* ------------------------------------------------------------------ */

int xdp_program(const uint8_t *data, size_t len) {
    /* Parse Ethernet */
    if (len < sizeof(struct eth_hdr))
        return XDP_DROP;
    struct eth_hdr *eth = (struct eth_hdr *)data;
    if (ntoh16(eth->eth_type) != 0x0800)
        return XDP_PASS;               /* non-IPv4: let through */

    /* Parse IPv4 */
    size_t off = sizeof(struct eth_hdr);
    if (len < off + 20)
        return XDP_DROP;
    struct ipv4_hdr *ip = (struct ipv4_hdr *)(data + off);
    int ihl = (ip->version_ihl & 0x0F) * 4;
    if (ihl < 20 || len < off + (size_t)ihl)
        return XDP_DROP;

    int      proto_idx;
    uint16_t src_port = 0, dst_port = 0;

    switch (ip->protocol) {
    case 6:                     /* TCP */
        proto_idx = 0;
        if (len >= off + (size_t)ihl + 4) {
            src_port = (data[off + ihl]     << 8) | data[off + ihl + 1];
            dst_port = (data[off + ihl + 2] << 8) | data[off + ihl + 3];
        }
        break;
    case 17:                    /* UDP */
        proto_idx = 1;
        if (len >= off + (size_t)ihl + 4) {
            src_port = (data[off + ihl]     << 8) | data[off + ihl + 1];
            dst_port = (data[off + ihl + 2] << 8) | data[off + ihl + 3];
        }
        break;
    case 1:                     /* ICMP */
        proto_idx = 2;
        break;
    default:
        proto_idx = 3;
    }

    /* Update protocol counter (BPF_MAP_TYPE_ARRAY increment) */
    __sync_fetch_and_add(&proto_counts[proto_idx], 1);

    /* Track per-flow bytes for TCP/UDP (BPF_MAP_TYPE_HASH semantics) */
    if (proto_idx < 2) {
        int found = 0;
        for (int i = 0; i < flow_cnt; i++) {
            if (flows[i].src_ip   == ip->src_ip  &&
                flows[i].dst_ip   == ip->dst_ip  &&
                flows[i].protocol == ip->protocol &&
                flows[i].src_port == src_port    &&
                flows[i].dst_port == dst_port) {
                flows[i].bytes += ntoh16(ip->total_length);
                found = 1;
                break;
            }
        }
        if (!found && flow_cnt < MAX_FLOWS) {
            flows[flow_cnt].src_ip   = ip->src_ip;
            flows[flow_cnt].dst_ip   = ip->dst_ip;
            flows[flow_cnt].protocol = ip->protocol;
            flows[flow_cnt].src_port = src_port;
            flows[flow_cnt].dst_port = dst_port;
            flows[flow_cnt].bytes    = ntoh16(ip->total_length);
            flow_cnt++;
        }
    }
    return XDP_PASS;
}

/* ------------------------------------------------------------------ */
/*  Test traffic generation                                           */
/* ------------------------------------------------------------------ */

static void build_tcp_pkt(uint8_t *buf, size_t len,
                          const uint8_t *sip, const uint8_t *dip,
                          uint16_t sp, uint16_t dp)
{
    memset(buf, 0, len);
    buf[12] = 0x08; buf[13] = 0x00;               /* EtherType IPv4 */
    buf[14] = 0x45;                               /* ver=4, IHL=5 */
    buf[16] = 0x00; buf[17] = 40;                 /* total_length=40 */
    buf[23] = 6;                                   /* protocol=TCP */
    memcpy(buf + 26, sip, 4);                      /* src IP */
    memcpy(buf + 30, dip, 4);                      /* dst IP */
    buf[34] = (sp >> 8) & 0xFF; buf[35] = sp & 0xFF;
    buf[36] = (dp >> 8) & 0xFF; buf[37] = dp & 0xFF;
}

static void build_udp_pkt(uint8_t *buf, size_t len,
                          const uint8_t *sip, const uint8_t *dip,
                          uint16_t sp, uint16_t dp)
{
    build_tcp_pkt(buf, len, sip, dip, sp, dp);
    buf[16] = 0x00; buf[17] = 30;                 /* smaller packet */
    buf[23] = 17;                                  /* protocol=UDP  */
}

static void build_icmp_pkt(uint8_t *buf, size_t len,
                           const uint8_t *sip, const uint8_t *dip)
{
    build_tcp_pkt(buf, len, sip, dip, 0, 0);
    buf[16] = 0x00; buf[17] = 28;                 /* total_length=28 */
    buf[23] = 1;                                   /* protocol=ICMP */
}

static void simulate_traffic(void) {
    uint8_t buf[64];
    uint8_t sipA[] = {192, 168, 1, 1};
    uint8_t dipB[] = {10,   0,   0, 1};
    uint8_t dns[]  = {8,    8,   8, 8};

    printf("  Processing packets through XDP program...\n");

    /* 20 TCP packets */
    build_tcp_pkt(buf, sizeof(buf), sipA, dipB, 80, 8080);
    for (int i = 0; i < 20; i++) xdp_program(buf, 54);
    printf("    TCP x20: XDP_PASS\n");

    /* 10 UDP DNS packets */
    build_udp_pkt(buf, sizeof(buf), sipA, dns, 53, 12345);
    for (int i = 0; i < 10; i++) xdp_program(buf, 54);
    printf("    UDP x10: XDP_PASS\n");

    /* 5 ICMP pings */
    build_icmp_pkt(buf, sizeof(buf), sipA, dns);
    for (int i = 0; i < 5; i++) xdp_program(buf, 54);
    printf("    ICMP  x5: XDP_PASS\n");
}

/* ------------------------------------------------------------------ */
/*  main                                                              */
/* ------------------------------------------------------------------ */

int main(void) {
    printf("=== Software-Defined Networking & eBPF ===\n\n");

    printf("SDN simulation: see python3 code/main.py\n");
    printf("- 3-switch tree topology\n");
    printf("- Learning switch (MAC learning via PACKET_IN)\n");
    printf("- Static flow rule installation via controller\n\n");

    printf("=== XDP Packet Counter (in-C eBPF simulation) ===\n\n");
    simulate_traffic();

    printf("\n  Protocol distribution (BPF_MAP_TYPE_ARRAY):\n");
    for (int i = 0; i < MAX_PROTO; i++) {
        int bar = proto_counts[i] > 20 ? 20 : (int)proto_counts[i];
        printf("    %-6s: %3llu pkts ", proto_names[i],
               (unsigned long long)proto_counts[i]);
        for (int j = 0; j < bar; j++) putchar('#');
        putchar('\n');
    }

    printf("\n  Top flows by byte volume:\n");
    int shown = 0;
    for (int i = 0; i < flow_cnt && shown < 5; i++) {
        uint32_t sip = ntoh32(flows[i].src_ip);
        uint32_t dip = ntoh32(flows[i].dst_ip);
        const char *pname = proto_names[
            flows[i].protocol == 6  ? 0 :
            flows[i].protocol == 17 ? 1 :
            flows[i].protocol == 1  ? 2 : 3];
        printf("    %d.%d.%d.%d:%-5d -> %d.%d.%d.%d:%-5d (%s): %llu bytes\n",
               (sip >> 24) & 0xFF, (sip >> 16) & 0xFF,
               (sip >> 8)  & 0xFF, sip & 0xFF, flows[i].src_port,
               (dip >> 24) & 0xFF, (dip >> 16) & 0xFF,
               (dip >> 8)  & 0xFF, dip & 0xFF, flows[i].dst_port,
               pname, (unsigned long long)flows[i].bytes);
        shown++;
    }
    putchar('\n');
    return 0;
}
